pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/PeerServerSegmentFetcher.java

```diff
@@ -0,0 +1,149 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.utils.fetcher;
+
+import java.io.File;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import org.apache.commons.configuration.Configuration;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixManager;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.pinot.common.utils.CommonConstants;
+import org.apache.pinot.common.utils.LLCSegmentName;
+import org.apache.pinot.common.utils.StringUtil;
+import org.apache.pinot.common.utils.helix.HelixHelper;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.plugin.PluginManager;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This segment fetcher downloads the segment by first discovering the server having the segment through external view
+ * of a Pinot table and then downloading the segment from the peer server using a configured http or https fetcher. By
+ * default, HttpSegmentFetcher is used.
+ * The format fo expected segment address uri is
+ *    server:///segment_name
+ * Note the host component is empty.
+ * To use this segment fetcher, servers need to put "server" in their segment fetcher protocol.
+ */
+public class PeerServerSegmentFetcher extends BaseSegmentFetcher {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PeerServerSegmentFetcher.class);
+  private static final String PEER_2_PEER_PROTOCOL = "server";
+  private static final String DOWNLOADER_CLASS = ".downloader.class";
+  private HelixManager _helixManager;
+  private String _helixClusterName;
+  private HttpSegmentFetcher _httpSegmentFetcher;
+
+  public PeerServerSegmentFetcher(Configuration config, HelixManager helixManager, String helixClusterName) {
+    _helixManager = helixManager;
+    _helixClusterName = helixClusterName;
+    String segmentDownloaderClass = config.getString(PEER_2_PEER_PROTOCOL + DOWNLOADER_CLASS);
+    try {
+      _httpSegmentFetcher = PluginManager.get().createInstance(segmentDownloaderClass);
```

Comment:
So, essentially we are saying that it can either be http or https segment fetcher (or some other class that extends from these), right? We need to get the config story right. Can we configure just the scheme and construct the rest of the url here and initialize the right fetcher? Do you anticipate any other class than these two being used?

Reply:
Because this fetcher download segments from peer servers, http and https segment fetchers would be sufficient for usage considering the download API server supports. 

What do you mean "construct the rest of the url here"?  In the constructor? The complete url is a function of segment and we do not have segment info here. Right now the server discovery and url construction is done in per segment fetch basis. @mcvsubbu

Reply:
I meant construct uri in this class (like you are doing) but based on the configured scheme

Reply:
The config I use now passed in the fetcher class (mainly for ease of testing). I can pass in scheme (http or https) too together with the class. But let us put this config choice in a broader discussion we had offline so that we plan everything in one discussion.